xezno/mocha
# (Draft) Implement Networking using GameNetworkingSockets and Quake 3-like snapshot system

Note - this is a draft pull request and isn't ready to be merged yet.

## Description
This pull request implements networking functionality for our game engine using the GameNetworkingSockets library for transport and a snapshot system similar to Quake 3's architecture. It includes the following features:

- GameNetworkingSockets transport for reliable and efficient data transfer
- Snapshot system for sending/receiving game state updates
- Prediction algorithms to minimize lag
- Lag compensation for more accurate gameplay
- Graceful handling of disconnections and reconnections
- Debugging and profiling tools to help diagnose networking issues

## Changes Made
To implement the networking functionality, the following changes were made to the networking branch:

- Added GameNetworkingSockets library as a dependency
- Created `NetworkingManager` class to handle networking functionality
- Implemented snapshot system for sending/receiving game state updates
- Developed prediction algorithms to minimize lag
- Added lag compensation for more accurate gameplay
- Implemented graceful handling of disconnections and reconnections
- Added debugging and profiling tools to help diagnose networking issues

Various low-level engine changes have also been made:
- Mocha now compiles down to a library (Mocha.Host.lib). A root is then created by either Mocha or MochaDedicatedServer
- Bindings are no longer global and are now bound to a Root instance. This is currently accessed as `Global.NativeEngine` (or just `NativeEngine` in most cases)
- Native globals are now stored inside a `Global` namespace
- Re-wrote the base walk controller to use Jolt physics for collision and clipping
- Added C# -> C++ callbacks (#44)
- Added integer console variables
- `HandleMap` now has functions for `Remove`, `RemoveAt`, and `Find`

## API Changes

### Networking

**Synchronisation**
To synchronise a property or field over the network:
```cs
[Sync] public string MyNetworkedString { get; set; }
[Sync] public string MyOtherNetworkedString;
```

**Method Stripping**
To strip methods:
```cs
[Event.Tick, ServerOnly]
public void ServerTick()
{
    DebugOverlay.ScreenText( "Server Tick..." );
}

[Event.Tick, ClientOnly]
public void ClientTick()
{
    DebugOverlay.ScreenText( "Client Tick..." );
}
```

### Callbacks

**Managed -> Native Callbacks**
To register a callback that can be called from C++:
```cs
var callbackId = CallbackDispatcher.RegisterCallback( () => Log.Trace( "Hello" ) );
// Then pass callbackId to a native method that accepts a callback
```

## How to Test
To test the networking functionality, do the following:

1. Clone this branch and build the game engine
2. Run the game engine and connect to a server or host a game
3. Play the game and observe the networking behavior

## Related Issues
This pull request is related to the following issues:
- Closes #52
- Closes #44
